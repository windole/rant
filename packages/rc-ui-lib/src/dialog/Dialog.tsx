import React, { useContext } from 'react';
import classnames from 'classnames';

import Popup from '../popup';
import Button from '../button';
import ActionBar from '../action-bar';

import { DialogProps, DialogStatic } from './PropsType';
import { addUnit, noop } from '../utils';
import { BORDER_TOP, BORDER_LEFT } from '../utils/constant';
import ConfigProviderContext from '../config-provider/ConfigProviderContext';

const Dialog: React.FC<DialogProps> = (props) => {
  const { prefixCls, createNamespace } = useContext(ConfigProviderContext);
  const [bem] = createNamespace('dialog', prefixCls);

  const {
    width,
    title,
    theme,
    visible,
    message,
    className,
    messageAlign,
    closeOnClickOverlay,
    onClickCloseIcon,
    ...others
  } = props;

  const renderTitle = () => {
    if (props.title) {
      return (
        <div
          className={classnames(
            bem('header', {
              isolated: !props.message && !props.children,
            }),
          )}
        >
          {title}
        </div>
      );
    }
    return null;
  };

  const renderContent = () => {
    if (props.children) {
      return <div className={classnames(bem('content'))}>{props.children}</div>;
    }
    if (message) {
      return (
        <div className={classnames(bem('content', { isolated: !title }))}>
          <div
            className={classnames(
              bem('message', {
                'has-title': title,
                [messageAlign]: messageAlign,
              }),
            )}
          >
            {message}
          </div>
        </div>
      );
    }

    return null;
  };

  const renderButtons = () => (
    <div className={classnames(BORDER_TOP, bem('footer'))}>
      {props.showCancelButton && (
        <Button
          size="large"
          text={props.cancelButtonText || '取消'}
          className={classnames(bem('cancel'))}
          style={{ color: props.cancelButtonColor }}
          loading={props.cancelProps?.loading}
          disabled={props.cancelProps?.disabled}
          onClick={props.cancelProps?.loading ? noop : props.onCancel}
        />
      )}
      {props.showConfirmButton && (
        <Button
          size="large"
          text={props.confirmButtonText || '确认'}
          className={classnames(bem('confirm'), { [BORDER_LEFT]: props.showCancelButton })}
          round={theme === 'round-button'}
          style={{ color: props.confirmButtonColor }}
          loading={props.confirmProps?.loading}
          disabled={props.confirmProps?.disabled}
          onClick={props.confirmProps?.loading ? noop : props.onConfirm}
        />
      )}
    </div>
  );

  const renderRoundButtons = () => (
    <ActionBar className={classnames(bem('footer'))}>
      {props.showCancelButton && (
        <ActionBar.Button
          type="warning"
          text={props.cancelButtonText || '取消'}
          className={classnames(bem('cancel'))}
          color={props.cancelButtonColor}
          loading={props.cancelProps?.loading}
          disabled={props.cancelProps?.disabled}
          onClick={props.cancelProps?.loading ? noop : props.onCancel}
        />
      )}
      {props.showConfirmButton && (
        <ActionBar.Button
          type="danger"
          text={props.confirmButtonText || '确认'}
          className={classnames(bem('confirm'))}
          color={props.confirmButtonColor}
          loading={props.confirmProps?.loading}
          disabled={props.confirmProps?.disabled}
          onClick={props.confirmProps?.loading ? noop : props.onConfirm}
        />
      )}
    </ActionBar>
  );

  const renderFooter = () => {
    if (props.footer) return props.footer;
    return props.theme === 'round-button' ? renderRoundButtons() : renderButtons();
  };

  return (
    <Popup
      {...others}
      visible={visible}
      className={classnames(bem([theme]), className)}
      style={{ width: addUnit(width) }}
      aria-labelledby={title || message}
      closeOnClickOverlay={closeOnClickOverlay}
      onOpen={props.onOpen}
      onOpened={props.onOpened}
      onClose={props.onClose}
      onClosed={props.onClosed}
    >
      {renderTitle()}
      {renderContent()}
      {renderFooter()}
    </Popup>
  );
};

Dialog.defaultProps = {
  transition: 'rc-dialog-bounce',
  showConfirmButton: true,
  closeOnPopstate: true,
};

export default Dialog as React.FC<DialogProps> & DialogStatic;
