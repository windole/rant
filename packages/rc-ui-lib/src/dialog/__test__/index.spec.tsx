import React from 'react';
import { mount, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Dialog from '..';
import Button from '../../button';

describe('Dialog', () => {
  let wrapper;
  afterEach(() => {
    wrapper.unmount();
    jest.clearAllMocks();
  });

  it('should change confirm button color when using confirm-button-color prop', () => {
    wrapper = mount(<Dialog visible confirmButtonColor="red" />);
    const confirmButton = wrapper.find('.rc-dialog__confirm').at(0);
    expect(confirmButton.getDOMNode().style.color).toEqual('red');
  });

  it('should change cancel button color when using cancel-button-color prop', () => {
    wrapper = mount(<Dialog visible showCancelButton cancelButtonColor="red" />);
    const cancelButton = wrapper.find('.rc-dialog__cancel').at(0);
    expect(cancelButton.getDOMNode().style.color).toEqual('red');
  });

  it('should render button text correctly', () => {
    wrapper = mount(
      <Dialog
        visible
        showCancelButton
        cancelButtonText="Custom Cancel"
        confirmButtonText="Custom Confirm"
      />,
    );
    expect(wrapper.find('.rc-dialog__footer').html()).toMatchSnapshot();
  });

  it('should render default slot correctly', () => {
    wrapper = mount(<Dialog visible message="Custom Message" />);
    expect(wrapper.find('.rc-dialog__message').text()).toEqual('Custom Message');
    expect(wrapper.find('.rc-dialog__content').html()).toMatchSnapshot();
  });

  it('should render title slot correctly', () => {
    wrapper = mount(<Dialog visible title="Custom Title" />);
    expect(wrapper.find('.rc-dialog__header').text()).toEqual('Custom Title');
    expect(toJson(wrapper.find('.rc-dialog__header'))).toMatchSnapshot();
  });

  it('should emit open event when show prop is set to true', async () => {
    const onOpen = jest.fn();
    wrapper = mount(<Dialog onOpen={onOpen} />);
    await wrapper.setProps({ visible: true });
    expect(onOpen).toHaveBeenCalledTimes(1);
  });

  it('should emit close event when show prop is set to false', async () => {
    const onClose = jest.fn();
    const onClickCloseIcon = jest.fn();
    wrapper = mount(<Dialog onClose={onClose} closeable onClickCloseIcon={onClickCloseIcon} />);
    wrapper.setProps({ visible: true });

    wrapper.find('i').simulate('click');
    wrapper.setProps({ visible: false });
    expect(onClose).toHaveBeenCalledTimes(1);
  });

  it('should update width when using width prop', async () => {
    wrapper = mount(<Dialog visible width={200} />);
    expect(wrapper.find('.rc-dialog').at(1).getDOMNode().style.width).toEqual('200px');
  });

  it('should render footer correctly', () => {
    wrapper = mount(<Dialog visible message="message" footer={<Button>Footer</Button>} />);
    expect(wrapper.find('.rc-dialog').at(1).find('.rc-button__text').text()).toEqual('Footer');
  });

  it('should correctly when showCancelButton prop is true', () => {
    wrapper = mount(<Dialog visible message="message" showCancelButton />);
    expect(
      wrapper.find('.rc-dialog').at(1).find('.rc-dialog__cancel').find('.rc-button__text').text(),
    ).toEqual('取消');
  });

  it('should correctly when theme prop is round-button', () => {
    wrapper = shallow(<Dialog visible message="message" theme="round-button" showCancelButton />);
    expect(wrapper.find('.rc-dialog__footer').find('ActionBarButton')).toHaveLength(2);
  });
});
