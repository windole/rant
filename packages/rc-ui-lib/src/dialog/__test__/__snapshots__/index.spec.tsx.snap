// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dialog should render button text correctly 1`] = `"<div class=\\"rc-hairline--top rc-dialog__footer\\"><button class=\\"rc-dialog__cancel rc-button rc-button--default rc-button--large\\"><div class=\\"rc-button__content\\"><span class=\\"rc-button__text\\">Custom Cancel</span></div></button><button class=\\"rc-dialog__confirm rc-hairline--left rc-button rc-button--default rc-button--large\\"><div class=\\"rc-button__content\\"><span class=\\"rc-button__text\\">Custom Confirm</span></div></button></div>"`;

exports[`Dialog should render default slot correctly 1`] = `"<div class=\\"rc-dialog__content rc-dialog__content--isolated\\"><div class=\\"rc-dialog__message\\">Custom Message</div></div>"`;

exports[`Dialog should render title slot correctly 1`] = `
<div
  className="rc-dialog__header rc-dialog__header--isolated"
>
  Custom Title
</div>
`;
